cmd/fpga_plugin/fpga_plugin.go

```diff
@@ -208,6 +213,10 @@ func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
 
 		notifier.Notify(devTree)
 
+		if dp.scanOnce {
```

Comment:
I could re-use this idea ;)

Reply:
`time.Sleep()` was used as a the simplest possible vehicle for periodic scans. Nowadays `time.Ticker` is the golden standard for this type of tasks :)

https://gobyexample.com/tickers

Reply:
@rojkov Are you suggesting to change Scan implementation to use time.Ticker?

Reply:
Yes. I guess we can start with the FPGA plugin and then use the same approach in all other filters.

Reply:
Here is how it would look (quite cumbersome in my opinion) if we use time.Ticker:
```
diff --git a/cmd/fpga_plugin/fpga_plugin.go b/cmd/fpga_plugin/fpga_plugin.go
index e8d106c..abdf7a5 100644
--- a/cmd/fpga_plugin/fpga_plugin.go
+++ b/cmd/fpga_plugin/fpga_plugin.go
@@ -48,6 +48,9 @@ const (
        // When the device's firmware crashes the driver reports these values
        unhealthyAfuID       = "ffffffffffffffffffffffffffffffff"
        unhealthyInterfaceID = "ffffffffffffffffffffffffffffffff"
+
+       // Frequency of device scans
+       scanFrequency = 5 * time.Second
 )
 
 type getDevTreeFunc func(devices []device) dpapi.DeviceTree
@@ -173,21 +176,29 @@ type devicePlugin struct {
        ignoreEmptyRegions bool
        annotationValue    string
 
-       scanOnce bool
+       scanTicker *time.Ticker
+       scanDone   chan bool
 }
 
 // newDevicePlugin returns new instance of devicePlugin
 func newDevicePlugin(mode string, rootPath string) (*devicePlugin, error) {
-       sysfsPathOPAE := path.Join(rootPath, sysfsDirectoryOPAE)
-       devfsPath := path.Join(rootPath, devfsDirectory)
-       if _, err := os.Stat(sysfsPathOPAE); os.IsNotExist(err) {
-               sysfsPathDFL := path.Join(rootPath, sysfsDirectoryDFL)
-               if _, err = os.Stat(sysfsPathDFL); os.IsNotExist(err) {
-                       return nil, fmt.Errorf("kernel driver is not loaded: neither %s nor %s sysfs entry exists", sysfsPathOPAE, sysfsPathDFL)
+       dp, err := func(mode string, rootPath string) (*devicePlugin, error) {
+               sysfsPathOPAE := path.Join(rootPath, sysfsDirectoryOPAE)
+               devfsPath := path.Join(rootPath, devfsDirectory)
+               if _, err := os.Stat(sysfsPathOPAE); os.IsNotExist(err) {
+                       sysfsPathDFL := path.Join(rootPath, sysfsDirectoryDFL)
+                       if _, err = os.Stat(sysfsPathDFL); os.IsNotExist(err) {
+                               return nil, fmt.Errorf("kernel driver is not loaded: neither %s nor %s sysfs entry exists", sysfsPathOPAE, sysfsPathDFL)
+                       }
+                       return newDevicePluginDFL(sysfsPathDFL, devfsPath, mode)
                }
-               return newDevicePluginDFL(sysfsPathDFL, devfsPath, mode)
-       }
-       return newDevicePluginOPAE(sysfsPathOPAE, devfsPath, mode)
+               return newDevicePluginOPAE(sysfsPathOPAE, devfsPath, mode)
+       }(mode, rootPath)
+
+       dp.scanTicker = time.NewTicker(scanFrequency)
+       dp.scanDone = make(chan bool, 1) // buffered as we may send to it before Scan will listen to it
+
+       return dp, err
 }
 
 func (dp *devicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error {
@@ -205,19 +216,31 @@ func (dp *devicePlugin) PostAllocate(response *pluginapi.AllocateResponse) error
 
 // Scan starts scanning FPGA devices on the host
 func (dp *devicePlugin) Scan(notifier dpapi.Notifier) error {
-       for {
+       defer dp.scanTicker.Stop()
+       scan := func() error {
                devTree, err := dp.scanFPGAs()
                if err != nil {
                        return err
                }
-
                notifier.Notify(devTree)
+               return nil
+       }
+       // scan at the start
+       err := scan()
+       if err != nil {
+               return err
+       }
 
-               if dp.scanOnce {
+       for {
+               select {
+               case <-dp.scanDone:
                        return nil
+               case <-dp.scanTicker.C:
+                       err := scan()
+                       if err != nil {
+                               return err
+                       }
                }
-
-               time.Sleep(5 * time.Second)
        }
 }
 
diff --git a/cmd/fpga_plugin/fpga_plugin_test.go b/cmd/fpga_plugin/fpga_plugin_test.go
index f466bfe..1c46aef 100644
--- a/cmd/fpga_plugin/fpga_plugin_test.go
+++ b/cmd/fpga_plugin/fpga_plugin_test.go
@@ -154,12 +154,15 @@ func TestNewDevicePlugin(t *testing.T) {
        }
 }
 
-// nemptyNotifier implements Notifier interface.
-type emptyNotifier struct {
+// fakeNotifier implements Notifier interface.
+type fakeNotifier struct {
+       scanDone chan bool
 }
 
-// Notify is an empty function to satisfy Notifier interface
-func (n *emptyNotifier) Notify(newDeviceTree dpapi.DeviceTree) {}
+// Notify stops plugin Scan
+func (n *fakeNotifier) Notify(newDeviceTree dpapi.DeviceTree) {
+       n.scanDone <- true
+}
 
 func TestScan(t *testing.T) {
        root, err := ioutil.TempDir("", "test_new_device_plugin")
@@ -203,7 +206,6 @@ func TestScan(t *testing.T) {
                },
        }
 
-       notifier := &emptyNotifier{}
        for _, tc := range tcases {
                t.Run(tc.name, func(t *testing.T) {
                        err = createTestDirs(devfs, tc.sysfs, tc.devfsdirs, tc.sysfsdirs, tc.sysfsfiles)
@@ -216,8 +218,8 @@ func TestScan(t *testing.T) {
                                t.Fatalf("failed to create a device plugin: %+v", err)
                        }
 
-                       plugin.scanOnce = true
-                       err = plugin.Scan(notifier)
+                       err = plugin.Scan(&fakeNotifier{plugin.scanDone})
+
                        if tc.expectedErr && err == nil {
                                t.Error("unexpected success")
                        }
~

```

Reply:
It would look much better if there is a way to trigger ticker at the start.

Reply:
@bart0sh did you see how QAT testing is done? Here the test case could check `scanFPGA` result.

Reply:
> It would look much better if there is a way to trigger ticker at the start.

You could leave the endless `for` loop as it is and just replace `time.Sleep()` with `select` having the case `case <-dp.scanTicker.C:` empty.

Reply:
@mythi can you point me out to the code or just briefly explain the idea?

@rojkov thank you! That did the trick. Please, review.

Reply:
@bart0sh https://github.com/intel/intel-device-plugins-for-kubernetes/blob/master/cmd/qat_plugin/dpdkdrv/dpdkdrv_test.go#L342:L350

Reply:
@mythi thanks. I wanted to test Scan API. Wrapping Scan internals into a function looks OK to me too. please review current implementation and let me know if it's OK or you like QAT approach better.

Reply:
This looks OK to me and it'd make more sense for me to follow that in QAT too (i.e., first move to `ticker` and then move to test Scan API using what you have in this PR.

Reply:
I noticed I cannot easily move to Scan API because QAT tests check deviceTree results which.